Src/DependencyCollector/Shared/ServiceBusDiagnosticsListener.cs

```diff
@@ -147,25 +142,31 @@ private void SetCommonProperties(string eventName, object eventPayload, Activity
                 }
             }
 
-            PropertyFetcher statusFetcher = this.GetOrCreatePropertyFetcher(eventName, StatusPropertyName);
-            telemetry.Success = (TaskStatus)statusFetcher.Fetch(eventPayload) == TaskStatus.RanToCompletion;
+            telemetry.Success = this.FetchPayloadProperty<TaskStatus>(eventName, StatusPropertyName, eventPayload) == TaskStatus.RanToCompletion;
+        }
+
+        private T FetchPayloadProperty<T>(string eventName, string propertyName, object payload)
+        {
+            Property property = new Property(eventName, propertyName);
+            var fetcher = this.propertyFetchers.GetOrAdd(property, prop => new PropertyFetcher(prop.PropertyName));
+            return (T)fetcher.Fetch(payload);
         }
 
-        private PropertyFetcher GetOrCreatePropertyFetcher(string eventName, string propertyName)
+        private struct Property
         {
-            if (!this.propertyFetchers.TryGetValue(eventName, out Dictionary<string, PropertyFetcher> fetchers))
+            public readonly string PropertyName;
+            private readonly string eventName;
+
+            public Property(string eventName, string propertyName)
             {
-                fetchers = new Dictionary<string, PropertyFetcher>();
-                this.propertyFetchers.Add(eventName, fetchers);
+                this.eventName = eventName;
+                this.PropertyName = propertyName;
             }
 
-            if (!fetchers.TryGetValue(propertyName, out PropertyFetcher fetcher))
+            public override int GetHashCode()
```

Comment:
You should also override Equals (the default Equals for structs involves reflection).
Perhaps also implement `IEquatable<Property>`
```C#
public override bool Equals(object other) => (other as Property)?.Equals(this) ?? false;
public bool Equals(Property other) => this.eventName == other.eventName && this.PropertyName == other.PropertyName;